torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -3890,7 +3890,18 @@ Tensor embedding_dense_double_backward(const Tensor & grad, const Tensor & indic
 }
 
 Tensor index_backward(Tensor zeros_like_self, const torch::List<c10::optional<Tensor>>& indices, const Tensor& grad) {
-  return at::_index_put_impl_(zeros_like_self, indices, grad, true, true);
+  c10::SmallVector<Tensor> inputs;
+  for (const c10::optional<Tensor> index : indices) {
+    if (index.has_value()) {
+      inputs.push_back(index.value());
+    }
+  }
+  inputs.push_back(zeros_like_self);
+  inputs.push_back(grad);
+
+  return areAnyTensorSubclassLike(inputs)
```

Comment:
Can we add an overload for `areAnyTensorSubclassLike(const torch::List<c10::optional<Tensor>>&)`? We want to avoid allocating a SmallVector here because it will make the fast-path even slower

Reply:
Adding the above overload leads to ambiguity when used as `areAnyTensorSubclassLike({input, grad})` as it matches both

`inline bool areAnyTensorSubclassLike(TensorList tensors)` and `inline bool areAnyTensorSubclassLike(const torch::List<c10::optional<Tensor>>& tensors)`.

List can be default constructed with `std::initializer_list` 

https://github.com/pytorch/pytorch/blob/f3e81f3eed9a6560f5591acf169c149c4eea9d6d/aten/src/ATen/core/List.h#L240-L251

Should we just give this a new name like `areAnyOptionalTensorSubclassLike`?

Reply:
Good point about the ambiguity. New name sounds reasonable.